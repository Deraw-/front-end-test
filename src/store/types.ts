// tslint:disable-next-line:no-empty-interface
export interface RootState {}

export interface Product extends Object {
	id: number;
	quantity: number;
}

export interface CartProduct extends Product {
	name: string;
	price: number;
	image: string;
}

// This is used to properly type components using mappers from Vuex
// see https://github.com/vuejs/vuex/pull/1121
// tslint:disable-next-line: no-empty-interface
export default interface VuexBindings {
	cartProducts: CartProduct[];
}
