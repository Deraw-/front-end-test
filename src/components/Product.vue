<template>
	<li
		class="product"
	>
		<img
			:src="`/api/img/${product.image}`"
			alt=""
			class="image"
		>

		<div class="product-desc">
			<h3 class="name">
				{{ product.name }}
			</h3>

			<p class="number">
				{{ $tc('components.product.number', product.number) }}
			</p>
		</div>

		<div class="actions">
			<span class="price">
				{{ product.price | euros }}
			</span>

			<UpdateCartBtn
				:product="product"
			/>
		</div>
	</li>
</template>

<script lang="ts">
	import Vue from 'vue';

	import UpdateCartBtn from '@/components/UpdateCartBtn.vue';

	export default Vue.extend({
		name: 'Poduct',
		components: {
			UpdateCartBtn
		},
		props: {
			product: {
				type: [Object, Array],
				required: true
			}
		}
	});
</script>

<style lang="scss" scoped>
	.product {
		width: 100%;
		margin: 0 auto;
		min-width: 250px;

		display: flex;
		flex-direction: column;

		box-shadow: 0px 2px 15px 0px rgba(0, 0, 0, .2);

		.image {
			height: 200px;
			object-fit: cover;
		}

		.product-desc {
			padding: 8px 16px;
		}

		.price {
			flex: 1 0 auto;
		}

		.name {
			font-size: 1.2rem;
			margin-bottom: 10px;
		}

		.number {
			font-size: .9rem;
		}
	}

	.actions {
		display: flex;
		align-items: center;
		justify-content: flex-end;

		padding: 8px 16px;
	}
</style>
