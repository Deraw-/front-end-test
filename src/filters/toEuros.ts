const toEuros = (value: number): string => {
	const price = `${value.toFixed(2)} â‚¬`;

	return price.replace('.', ',');
};

export default toEuros;
