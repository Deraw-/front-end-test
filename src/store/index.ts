import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import VuexPersistence from 'vuex-persist';
import { RootState } from './types';

// Modules
import cart from './modules/cart';
import products from './modules/products';

// The store is saved in the browser's session
const vuexLocal = new VuexPersistence({
	storage: window.sessionStorage
});

Vue.use(Vuex);

// See https://vuex.vuejs.org/fr/getting-started.html for help
const store: StoreOptions<RootState> = {
	strict: true,
	state: {},
	// See https://vuex.vuejs.org/guide/modules.html for more info on modules
	modules: {
		cart,
		products
	},
	plugins: [
		vuexLocal.plugin
	]
};

export default new Vuex.Store<RootState>(store);
