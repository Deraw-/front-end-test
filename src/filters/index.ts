import Vue from 'vue';

import toEuros from './toEuros';

Vue.filter('euros', toEuros);
