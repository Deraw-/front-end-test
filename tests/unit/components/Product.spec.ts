import { shallowMount } from '@vue/test-utils';
import Product from '@/components/Product.vue';

import localVue from 'tests-unit/index';
import store from '@/store/';
import '@/filters/';

// Data
import { CartProduct } from '@/store/types';
const product: CartProduct = {
	id: 0,
	quantity: 15,
	name: 'name',
	price: 15,
	image: '/img.jpg'
};

// Build component before each test
let wrapper: any;
beforeEach(() => {
	wrapper = shallowMount(Product, {
		localVue,
		store,
		propsData: {
			product
		}
	});
});

// Tests
describe('Product.vue', () => {
	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
