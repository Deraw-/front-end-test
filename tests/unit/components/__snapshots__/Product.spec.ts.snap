// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product.vue renders correctly 1`] = `
<li class="product"><img src="/api/img//img.jpg" alt="" class="image">
  <div class="product-desc">
    <h3 class="name">
      name
    </h3>
    <p class="number">
      components.product.number
    </p>
  </div>
  <div class="actions"><span class="price">
			15,00 â‚¬
		</span>
    <updatecartbtn-stub product="[object Object]"></updatecartbtn-stub>
  </div>
</li>
`;
